torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -1156,32 +1155,34 @@ def clip_grad_norm_(
                 "`clip_grad_norm_()` should only be called on the root FSDP instance"
             )
         self._assert_state(TrainingState.IDLE)
-        _wait_for_computation_stream(
-            torch.cuda.current_stream(),
-            self._streams["unshard"],
-            self._streams["pre_unshard"],
-        )
         max_norm = float(max_norm)
         norm_type = float(norm_type)
         # Perform local gradient norm computation, where sharded and
         # non-sharded parameters must be handled separately
         sharded_params = set()
         nonsharded_params = set()  # `NO_SHARD` or not FSDP-managed
+        grads: List[torch.Tensor] = []
         for handle in FullyShardedDataParallel._fsdp_handles(self):
             target_set = (
                 sharded_params if handle.uses_sharded_strategy else nonsharded_params
             )
             if handle._use_orig_params:
                 for param in handle.flat_param._params:
                     target_set.add(param)
+                    if param.grad is not None:
+                        grads.append(param.grad)
             else:
                 target_set.add(handle.flat_param)
+                if handle.flat_param.grad is not None:
+                    grads.append(handle.flat_param.grad)
         for param in self.parameters():
             not_fsdp_managed = (
                 param not in sharded_params and param not in nonsharded_params
             )
             if not_fsdp_managed:
                 nonsharded_params.add(param)
+                if param.grad is not None:
+                    grads.append(param.grad)
         local_sharded_norm = _get_grad_norm(sharded_params, norm_type).to(
             self.compute_device
         )
```

Comment:
looks like the `_get_grad_norm` uses grad values. Where do we make sure that reduce scatter on grad is done?

Reply:
Post-backward final callback:
https://github.com/pytorch/pytorch/blob/c599cf24ad01a4befa58642f9e703792f53b79fa/torch/distributed/fsdp/_runtime_utils.py#L701-L702